### PR TITLE
Support ProDiff in TTS

### PR DESCRIPTION
**Pull request by Fhrozen**
This draft has the filtered codes for a working ProDiff.
Using as reference: https://github.com/Rongjiehuang/ProDiff
ProDiff: https://arxiv.org/pdf/2207.06389.pdf

Some samples of a model trained with MFA aligments + HiFiGan Decoding.
[samples.zip](https://github.com/espnet/espnet/files/10160263/samples.zip)

### CODE REVIEW

**Comment by Fhrozen** [comment]
@kan-bayashi 
I suppose this is next?

**Review by kan-bayashi — COMMENTED** [review]
Could you add some unit test?

**Comment by Fhrozen** [comment]
@kan-bayashi test added. Pls leave any comment if anything else required..

File: espnet2/tts/prodiff/prodiff.py [hunk 1]
```diff
@@ -0,0 +1,837 @@
...
+    ):
+        """Reset parameters of the model.
+
+        Args:
+            init_type (str): Type of initialization.
+            init_enc_alpha (float): Value of the initialization for the encoder.
+            init_dec_alpha (float): Value of the initialization for the decoder.
+
+        Returns:
+            None
```
Comment (kan-bayashi) [inline espnet2/tts/prodiff/prodiff.py:826]:
```suggestion
```

File: test/espnet2/tts/test_prodiff.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    spk_embed_dim,
+    spk_embed_integration_type,
+    diffusion_scheduler,
+    encoder_type,
+    use_gst,
+    spks,
+    langs,
+):
+    if not is_torch_1_7_plus:
+        return None
```
Comment (kan-bayashi) [inline test/espnet2/tts/test_prodiff.py:33]:
```suggestion
@pytest.mark.skipif(
    is_torch_1_7_plus,
    reason="Pytorch >= 1.7 is required.",
)
def test_prodiff(
    reduction_factor,
    spk_embed_dim,
    spk_embed_integration_type,
    diffusion_scheduler,
    encoder_type,
    use_gst,
    spks,
    langs,
):
```

File: test/espnet2/tts/test_prodiff.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        inputs.update(durations=torch.tensor([2, 2, 0], dtype=torch.long))
+        inputs.update(pitch=torch.tensor([2, 2, 0], dtype=torch.float).unsqueeze(-1))
+        inputs.update(energy=torch.tensor([2, 2, 0], dtype=torch.float).unsqueeze(-1))
+        model.inference(**inputs, use_teacher_forcing=True)
+
+
+@pytest.mark.parametrize("reduction_type", ["none", "mean"])
+def test_ssim(reduction_type):
+    if not is_torch_1_7_plus:
+        return None
```
Comment (kan-bayashi) [inline test/espnet2/tts/test_prodiff.py:129]:
```suggestion
@pytest.mark.parametrize("reduction_type", ["none", "mean"])
@pytest.mark.skipif(
    is_torch_1_7_plus,
    reason="Pytorch >= 1.7 is required.",
)
def test_ssim(reduction_type):
```

File: egs2/ljspeech/tts1/conf/tuning/train_prodiff.yaml [hunk 4]
```diff
@@ -0,0 +1,98 @@
+# This configuration is for ESPnet2 to train ProDiff.
+# Train in 2 GPUs (RTX3060) for 2 days.
+
```
Comment (kan-bayashi) [inline egs2/ljspeech/tts1/conf/tuning/train_prodiff.yaml:3]:
```suggestion
# This configuration is for ESPnet2 to train ProDiff.
# Train in 2 GPUs (RTX3060) for 2 days.
```

File: egs2/ljspeech/tts1/conf/tuning/train_prodiff.yaml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
```
Comment (kan-bayashi) [inline egs2/ljspeech/tts1/conf/tuning/train_prodiff.yaml:4]:
```suggestion
# This configuration is for ESPnet2 to train ProDiff.
# Train in 2 GPUs (RTX3060) for 2 days.
```

File: espnet2/tts/prodiff/denoiser.py [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+    ) -> torch.Tensor:
+        """Calculate forward for denoising diffusion.
+
+        Args:
+            xs_noisy (torch.Tensor): Input tensor.
+            step (torch.Tensor): Number of step.
+            condition (torch.Tensor): Conditioning tensor.
+
+        Returns:
+            torch.Tensor: Denoised tensor.
```
Comment (kan-bayashi) [inline espnet2/tts/prodiff/denoiser.py:254]:
```suggestion
        Returns:
            torch.Tensor: Denoised tensor.

```

**Review by kan-bayashi — APPROVED** [review]
Thanks @Fhrozen, this is really cool :)
I slightly modified docstring format and skip condition in pytest.

**Comment by candlewill** [comment]
@Fhrozen I've been using Prodiff for other language tasks recently, but the synthesized audio output has some issues. If I don't use `--vocoder_file`, it will use the Griffin-Lim vocoder, which produces low-quality audio. If I want to specify using a neural vocoder with `--vocoder_file`, which pre-trained hifigan vocoder should I use?

**Comment by Fhrozen** [comment]
@candlewill I tried with hifigan vctkin japanese and it was pretty good. Now, you will need to be careful with some feats of the vocoder: sample rate, hop size.

**Comment by candlewill** [comment]
Thank you for your reply. Finally, I figured out the reason. When using a pre-trained vocoder model, the `stats.h5` file also needs to be downloaded, otherwise, normalization will be problematic.
